### PR TITLE
SecretnessAnalysis and ConvertIfToSelect fixes

### LINKED ISSUE #839: Secretness Analysis is potentially incorrect for ops that are not `IsolatedFromAbove`
I think that, as written, the analysis might only be correct for operations that are `IsolatedFromAbove` (i.e., do not use SSA values from above/outside their own region), which the `scf` operations aren't:

```llvm
%p = ... : i32 // something public
%s = ... : i32 // something that's (potentially only "transitively") secret
%r = scf.if(%cond : i1) -> i32 { // assume a public condition
    %sum = arith.addi %s, %s : i32 // secret
    %scf.yield %sum : i32 // also secret
} else {
    %scf.yield %p : i32 // doesn't really matter for this example
}
%ss = arith.addi %r, %r : i32 // should ALSO be secret, but won't be marked as such?
```

### PR DESCRIPTION
**Pull request by AlexanderViand-Intel**
Fixes #839 by conservatively treating ops with regions (except control-flow ops, which the analysis framework handles directly) as "unknown" secretness (except if they have secret operands, in which case they're already "secret" anyway). See #888 for how this could be made a bit less conservative (not a high priority though, imho).

Investigating that issue also brought up some crashes in `ConvertIfToSelect`, because (a) it did not update the secretness of the operations it created and (b) it did not consider that a secretness could be absent or uninitialized. This PR fixes both issues. I guess this might also be an issue in the other data-oblivious transformations (@MeronZerihun can you please check that?)

Finally, this PR also adds a convenient (DEBUG-only) "printer" for the secretness analysis to `--convert-if-to-select`, which should make further fixes/improvements a bit easier. This works in the same way as the TargetSlot analysis, i.e., walking the IR and adding attributes. I wonder if there's a better place to put this debug logic, like maybe a debug-only pass that does nothing but run and the secretness analysis and augment the IR (@j2kun?)

### CODE REVIEW

**Comment by j2kun** [comment]
> I wonder if there's a better place to put this debug logic, like maybe a debug-only pass that does nothing but run and the secretness analysis and augment the IR.

That sounds fair. I think there's probably a further step we could take, adding a util function like

```cpp
template <typename LatticeTy>
void attachLatticeAsAttribute(
    MLIRContext *context, 
    Operation *rootOp, 
    const std::string& attrName, 
    SomeCallbackType setAttr) {
  OpBuilder builder(context);
  rootOp->walk([&](Operation *op) {
    if (op->getNumResults() == 0) return;
    auto *lattice =
        solver.lookupState<LatticeTy>(op->getResult(0));
    if (!lattice) {
      op->setAttr(attrName, builder.getStringAttr("null"));
      return;
    }
    if (!lattice->getValue().isInitialized()) {
      op->setAttr(attrName, builder.getStringAttr("unknown"));
      return;
    }
    setAttr(op, lattice);
  });

  llvm::dbg() << "Op after attaching lattice attributes: " << *rootOp;
}
```

**Review by j2kun — APPROVED** [review]

File: lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp [hunk 1]
```diff
@@ -50,15 +54,16 @@ void SecretnessAnalysis::visitOperation(
...
 
-  if (resultSecretness.isInitialized()) {
-    // Uninitialized operand: theoretically this should not happen, but it's
-    // easy to catch it this way if it does
-    if (isUninitializedOpFound && !resultSecretness.getSecretness()) {
+  // Uninitialized operand: "false" needs to be reverted to "unknown"
+  // "secret" can remain, as "unknown + secret = secret"
+  // As region-bearing ops are not yet supported in the secretness analysis
+  // (except for control-flow, which the analysis framework handles directly),
+  // we appply the same conservative logic if any regions are detected
```
Comment (MeronZerihun) [inline lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp:61]: Not sure how this wasn't caught by the codespell check:
```suggestion
  // we apply the same conservative logic if any regions are detected
```

**Review by MeronZerihun — CHANGES_REQUESTED** [review]
LGTM :)

File: lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp [hunk 2]
```diff
@@ -39,6 +39,10 @@ void SecretnessAnalysis::visitOperation(
   auto resultSecretness = Secretness();
   bool isUninitializedOpFound = false;
```
Comment (MeronZerihun) [inline lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp:41]:
```suggestion
// Propagate secretness for constants
```

**Review by MeronZerihun — APPROVED** [review]

**Comment by MeronZerihun** [comment]
> This PR fixes both issues. I guess this might also be an issue in the other data-oblivious transformations (@MeronZerihun can you please check that?)

I think the other transformations will also have similar issues; I will make sure to replicate the fix and create a PR.

**Review by asraa — APPROVED** [review]
Thanks! I'll merge with the comment fixes

**Comment by AlexanderViand-Intel** [comment]
> Thanks! I'll merge with the comment fixes

Fixes pushed & rebased

**Pull request merged**